provider/balance_checker.go

```diff
@@ -2,153 +2,316 @@ package provider
+	resp := leaseCheckResponse{
+		lid:   lid,
+		state: respStateNextCheck,
 	}
 
-	balance := result.Balance.Amount
-	balanceGauge.Set(float64(balance.Uint64()))
+	if scheduledWithdraw {
+		resp.state = respStateScheduledWithdraw
+	}
 
-	if bc.cfg.MinimumBalanceThreshold == 0 {
-		return false, nil
+	var syncInfo *tmrpc.SyncInfo
+	syncInfo, resp.err = bc.session.Client().NodeSyncInfo(ctx)
+	if resp.err != nil {
+		return resp
 	}
 
-	tooLow := sdk.NewIntFromUint64(bc.cfg.MinimumBalanceThreshold).GT(balance)
+	if syncInfo.CatchingUp {
+		resp.err = aclient.ErrNodeNotSynced
+		return resp
+	}
 
-	return tooLow, nil
+	var dResp *dtypes.QueryDeploymentResponse
+	var lResp *mtypes.QueryLeasesResponse
+
+	// Fetch the balance of the escrow account
+	dResp, resp.err = bc.aqc.Deployment(ctx, &dtypes.QueryDeploymentRequest{
+		ID: lid.DeploymentID(),
+	})
+
+	if resp.err != nil {
+		return resp
+	}
+
+	lResp, resp.err = bc.aqc.Leases(ctx, &mtypes.QueryLeasesRequest{
+		Filters: mtypes.LeaseFilters{
+			Owner: lid.Owner,
+			DSeq:  lid.DSeq,
+			State: "active",
+		},
+	})
+
+	if resp.err != nil {
+		return resp
+	}
+
+	totalLeaseAmount := sdk.NewDec(0)
+	for _, lease := range lResp.Leases {
+		totalLeaseAmount = totalLeaseAmount.Add(lease.Lease.Price.Amount)
+	}
+
+	balanceRemain := util.LeaseCalcBalanceRemain(dResp.EscrowAccount.TotalBalance().Amount,
+		syncInfo.LatestBlockHeight,
+		dResp.EscrowAccount.SettledAt,
+		totalLeaseAmount)
+
+	blocksRemain := util.LeaseCalcBlocksRemain(balanceRemain, totalLeaseAmount)
+
+	// lease is out of funds
+	if blocksRemain <= 0 {
+		resp.state = respStateOutOfFunds
+		resp.checkAfter = time.Minute * 10
+	} else {
+		blocksPerCheckInterval := int64(bc.cfg.LeaseFundsCheckInterval / netutil.AverageBlockTime)
+		if blocksRemain > blocksPerCheckInterval {
+			blocksRemain = blocksPerCheckInterval
+		}
+
+		resp.checkAfter = time.Duration(blocksRemain) * netutil.AverageBlockTime
+	}
+
+	return resp
 }
 
-func (bc *balanceChecker) startWithdrawAll() error {
-	return bc.bus.Publish(event.LeaseWithdrawNow{})
+func (bc *balanceChecker) startWithdraw(lid mtypes.LeaseID) error {
+	return bc.bus.Publish(event.LeaseWithdraw{LeaseID: lid})
 }
 
-func (bc *balanceChecker) run() {
-	defer bc.lc.ShutdownCompleted()
-	ctx, cancel := context.WithCancel(context.Background())
+func (bc *balanceChecker) run(startCh chan<- error) {
+	ctx, cancel := context.WithCancel(bc.ctx)
+
+	defer func() {
+		cancel()
+		bc.lc.ShutdownCompleted()
 
-	tick := time.NewTicker(bc.cfg.PollingPeriod)
-	withdrawalTicker := time.NewTicker(bc.cfg.WithdrawalPeriod)
+		for _, lState := range bc.leases {
+			if lState.tm != nil && !lState.tm.Stop() {
+				<-lState.tm.C
+			}
+		}
+	}()
 
-	var balanceCheckResult <-chan runner.Result
-	var withdrawAllResult <-chan runner.Result
+	leaseCheckCh := make(chan leaseCheckResponse, 1)
+	resultCh := make(chan runner.Result, 1)
+
+	subscriber, err := bc.bus.Subscribe()
```

Comment:
So it looks like `startCh` is used to send back a possible error from this? The way I got around this constraint was just by doing the call to `bus.Subscribe()` in `newBalanceChecker` and passing the subscriber as an argument to this function.